### DISCUSSION TITLE
Supported Upgrade Paths

### DISCUSSION THREAD

**Discussion by plumbis** [crossplane/crossplane#4569] [General]
In #4557 a user had issues upgrading from v1.7.0 to v1.13.2. 

What is our "supported" upgrade path? 

How far back do we want to support upgrades from/to? 

What should have been the (generic) recommendation for this user? 

How should we capture the API changes that would result in [this kind of behavior](https://github.com/crossplane/crossplane/issues/4557#issuecomment-1705199048)?

**Comment by dee0sap** [comment]
Btw is a user, I would appreciate it if any requirements were enforced in code.    That is if I attempt an unsupported upgrade then I would like a clear error message informing me of this written somewhere where I am most likely to see it.

**Comment by phisco** [comment]
Given that whatever we end up officially supporting, we should also be automatically testing. I'd say we should only officially support upgrading one minor version at a time, or at most skipping one minor, given that we officially support 3 minor releases at a time.

So:
- 1.x.y => 1.<x+1>.z: yes
- 1.x.y => 1.<x+2>.z: yes
- other cases: no

Regarding clearly erroring out on unsupported version bumps, this could be definitely done by the helm chart itself. Doing that from Crossplane itself would require having the version stored somewhere, which I don't think we have at the moment, but would be feasible. However, on one side I agree we should warn users they are doing something we don't officially support, but I would not fully prevent them from doing so if they are willing to assume such risk. So I would at least add a way to bypass it, if we decided to prevent that by default (which I would vote against).

**Reply by dee0sap** [reply]
Maybe a 'strict upgrade check' option that users could turn on?

Fwiw, I am pretty sure on my team we would turn that on just to we didn't unwittingly roll out something to production only to find out after the fact that some problem was due to an unsupported upgrade path being used.     A particular concern since, for better or worse, the task of rolling out new versions of software often gets assigned to the most junior people and they more likely to make such a mistake.

**Comment by jbw976** [comment]
A formal API promotion process has been defined in https://github.com/crossplane/crossplane/pull/6226 that allows for upgrades and downgrades in Crossplane when performed sequentially one minor version at a time. With the way that CRDs and custom resources are stored and upgraded over time, this is the best path to take to maintain safety for both upgrades and downgrades.

The API promotion guide can be viewed at https://github.com/crossplane/crossplane/blob/main/contributing/guide-api-promotion.md. It is also part of the PR checklist and Crossplane contributors should follow this guide diligently when promoting Crossplane APIs.

With that in mind, I'd say that Crossplane's upgrade (and downgrade) policy should be to sequentially perform them one minor version at a time, e.g. `v1.18` --> `v1.19` --> `v1.20` --> `v1.19` --> `v1.18`. The latest patch release on each of those minor versions should also be used, to ensure you have any fixes affecting upgrade/downgrade reliability for each minor version.
